pdp-server/src/api/authz/forward_to_opa.rs

```diff
@@ -0,0 +1,177 @@
+    body: &B,
+) -> Result<R, ForwardingError> {
+    // Create a new OPA request
+    let request = create_opa_request(body, state.settings.debug)?;
+
+    // Send the request to OPA
+    let response: OpaResponse<R> = send_raw_request_to_opa(state, endpoint, &request).await?;
+    Ok(response.result)
+}
+
+/// Generic function to forward requests to OPA
+async fn send_raw_request_to_opa<B: Serialize, R: DeserializeOwned>(
+    state: &AppState,
+    endpoint: &str,
+    body: &OpaRequest<B>,
+) -> Result<OpaResponse<R>, ForwardingError> {
+    // Build the OPA URL from the settings
+    let endpoint = endpoint.strip_prefix("/").unwrap_or(endpoint);
+    let opa_url = format!("{}/{}", state.settings.opa_url, endpoint);
+    debug!("Forwarding request to OPA at: {}", opa_url);
+
+    // Send request to OPA using the specialized OPA client
+    // TODO add timeout
+    let response = state.opa_client.post(&opa_url).json(body).send().await?;
+
+    // Check if the request was successful
+    if !response.status().is_success() {
+        let status = response.status();
+        return Err(ForwardingError::InvalidStatus(status));
+    }
+
+    // Parse the response body
+    let body = response.bytes().await?;
+    Ok(serde_json::from_slice(&body)?)
+}
+
+/// Helper function to create an OPA request from any serializable type
+///
+/// It serializes the value inside the 'input' field of the OPA request.
+/// If the value is an object, it injects a `use_debugger` field if the debug flag is set.
+fn create_opa_request<T: Serialize>(
+    value: T,
+    debug: Option<bool>,
+) -> Result<OpaRequest<serde_json::Value>, ForwardingError> {
+    let mut value = serde_json::to_value(&value)?;
+    if let Some(debug) = debug {
+        // Inject `use_debugger` if the value is an object
+        if let Object(ref mut obj) = value {
+            obj.insert("use_debugger".to_string(), Bool(debug));
+        }
+    }
+    Ok(OpaRequest { input: value })
+}
+
+/// A generic wrapper for OPA requests, wrapping the input data.
+/// https://www.openpolicyagent.org/docs/latest/integration/#named-policy-decisions
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct OpaRequest<T> {
+    pub input: T,
+}
+
+/// A generic wrapper for OPA responses, wrapping the result data.
+/// https://www.openpolicyagent.org/docs/latest/integration/#named-policy-decisions
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct OpaResponse<T> {
+    pub result: T,
+}
+
+#[derive(Debug, Error)]
+pub(crate) enum ForwardingError {
+    #[error("Failed to build request to OPA: {0}")]
+    BuildError(#[from] InvalidHeaderValue),
+    #[error("Failed to send request to OPA: {0}")]
+    RequestError(#[from] reqwest::Error),
+    #[error("OPA request failed with status: {0}")]
+    InvalidStatus(StatusCode),
+    #[error("Failed to parse OPA response: {0}")]
+    ParseError(#[from] serde_json::Error),
+}
+
+impl From<ForwardingError> for ApiError {
```

Comment:
ApiError is implemented elsewhere, I think it's better to implement the Into<ApiError> for ForwardingError

A developer would look for implementation near the struct and not elsewhere